apollo-router/src/axum_factory/listeners.rs

```diff
@@ -217,9 +218,55 @@ pub(super) async fn get_extra_listeners(
     Ok(listeners_and_routers)
 }
 
+// This macro unifies the logic tht deals with connections.
+// Ideally this would be a function, but the generics proved too difficult to figure out.
+macro_rules! handle_connection {
+    ($connection:expr, $connection_handle:expr, $connection_shutdown:expr, $connection_shutdown_timeout:expr, $received_first_request:expr) => {
+        let connection = $connection;
+        let mut connection_handle = $connection_handle;
+        let connection_shutdown = $connection_shutdown;
+        let connection_shutdown_timeout = $connection_shutdown_timeout;
+        let received_first_request = $received_first_request;
+        tokio::pin!(connection);
+        tokio::select! {
+            // the connection finished first
+            _res = &mut connection => {
+            }
+            // the shutdown receiver was triggered first,
+            // so we tell the connection to do a graceful shutdown
+            // on the next request, then we wait for it to finish
+            _ = connection_shutdown.notified() => {
+                connection_handle.shutdown();
+                connection.as_mut().graceful_shutdown();
+                // if the connection was idle and we never received the first request,
+                // hyper's graceful shutdown would wait indefinitely, so instead we
+                // close the connection right away
+                if received_first_request.load(Ordering::Relaxed) {
+                    // The connection may still not shutdown so we aplly a timeout from the configuration
+                    // Connections stuck terminating will keep the pipeline and everything related to that pipeline
+                    // in memory.
+
+                    if let Err(_) = connection.timeout(connection_shutdown_timeout).await {
+                        tracing::warn!(
+                            timeout = connection_shutdown_timeout.as_millis() / 1000,
+                            server.address = connection_handle.connection_ref.address.to_string(),
+                            schema.id = connection_handle.connection_ref.pipeline_ref.schema_id,
+                            config.hash = connection_handle.connection_ref.pipeline_ref.config_hash,
+                            launch.id = connection_handle.connection_ref.pipeline_ref.launch_id,
+                            "connection shutdown exceeded, forcing close",
+                        );
+                    }
+                }
+            }
+        }
+    };
+}
+
+#[allow(clippy::too_many_arguments)]
```

Comment:
not a fan of this, but i hear you, it's difficult to do anything else reasonably

Reply:
Agree. This needs a larger refactoring though.